lib/ansible/modules/windows/win_domain_user.ps1

```diff
@@ -4,6 +4,96 @@
 
 #Requires -Module Ansible.ModuleUtils.Legacy
 
+Function Test-Credential {
+    param(
+        [String]$Username,
+        [String]$Password,
+        [String]$Domain = $null
+    )
+    if (($Username.ToCharArray()) -contains [char]'@') {
+        # UserPrincipalName
+        $Domain = $null # force $Domain to be null, to prevent undefined behaviour, as a domain name is already included in the username
+    } elseif (($Username.ToCharArray()) -contains [char]'\') {
+        # Pre Win2k Account Name
+        $Username = ($Username -split '\')[0]
+        $Domain = ($Username -split '\')[1]
+    } else {
+        # No domain provided, so maybe local user, or domain specified separately.
+    }
+
+    # More information about LogonUser at https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-logonusera
+    $platform_util = @'
```

Comment:
I've just recently added the `Ansible.AccessToken` module util which can be used here. Just add `#AnsibleRequires -CSharpUtil Ansible.AccessToken` to the top of your script and then this can be simplified with;

```
try {
    $handle = [Ansible.AccessToken.TokenUtil]::LogonUser($Username, $Domain, $Password, "Network", "Default")
    $handle.Dispose()
    return $true
} catch [Ansible.AccessToken.Win32Exception] {
    # following errors indicate the creds are correct but the user was
    # unable to log on for other reasons, which we don't care about
    $success_codes = @(
        0x0000052F,  # ERROR_ACCOUNT_RESTRICTION
        0x00000530,  # ERROR_INVALID_LOGON_HOURS
        0x00000531,  # ERROR_INVALID_WORKSTATION
        0x00000569  # ERROR_LOGON_TYPE_GRANTED
    )
    $failed_codes = @(
        0x0000052E,  # ERROR_LOGON_FAILURE
        0x00000532  # ERROR_PASSWORD_EXPIRED
    )

    if ($_.Exception.NativeErrorCode -in $failed_codes) {
        return $false
    } elseif ($_.Exception.NativeErrorCode -in $success_codes) {
        return $true
    } else {
        # an unknown failure, reraise exception
        throw $_
    }
}
```

This way you don't need any of the embedded C# code and no need for the weird `Add-Type` workaround with temp files.